squiRL/common/agents.py

```diff
@@ -87,13 +82,13 @@ def play_step(
             done (bool): Indicates if a step is terminal
         """
         action = self.get_action(net)
+        action = [action] if type(action) == int else action
 
         # do step in the environment
-        new_obs, reward, done, _ = self.env.step(action)
-        exp = Experience(self.obs, action, reward, done, new_obs)
+        self.env.act(action)
+        reward, new_obs, first = self.env.observe()
+        exp = Experience(self.obs, action, reward, first, new_obs)
```

Comment:
doesn't observe return a tuple of lists? This means that when you create Experience, you have multiple obs saved. Shouldn't this be split in as many Experiences as environments?

Reply:
Yes, but why would we split them? If we split and add them as separate experiences it would be difficult to disentangle the episodes gathered by the different envs from each other. Each column is an env. The entire column is the full episode from that env.

Lemme know your thoughts on this.

Reply:
The benefit of saving as full episodes is that you won't have any complicated collate functions afterward. One way to do this is to have a different storage instead of deque. For example, you could have a list where each entry is associated with an episode. Then, each parallel environment saves to it's own index in that list. When an episode finishes, a new index is given to that environment (all of this is done inside the storage class). Once you've hit x full episodes in your storage, you just yield them. This also gives you better control over how you want the sampling to be done (x episodes with a minimum of 1 per environment, or first x episodes sampled).

Reply:
I followed your advice. Now data is saved as full episodes.